Geometric interpretation on plugs

mattigruener opened the pull request:
Hey John,

I assume this will need a bit of back and forth between us, but I wanted to show you a prototype so you can tell me what you want changed.

The need for having a geometric interpretation on `V3fPlugs` came up when I started using the `PlugAdder`. I would like to react differently to connections made to plugs representing normals and points etc (basically to create the respective `oslOut<Type>` nodes). We also always had the problem of default values on plugs representing normals not coming through properly. This PR should address both of these issues. A remaining problem is that Arnold doesn't have a data type for normals... Converting the normals  to a color and showing it as Cs in the viewport works, but previewing the `VectorToColor` conversion doesn't.

As for implementation details - I'm not sure whether you like that I added an additional constructor for the `CompoundNumericPlug`? We decided to do that because we had the impression we'd want to keep the flags as the last arg? If that's not the case, we could save the string mangling in the serialiser and use that extraArg mechanism instead to append the geometric interpretation as an arg? If we need that constructor (and Dan and me both felt that having the interpretation very early in the arg list kinda makes sense because it kinda determines the type of the plug), - do we want to use an init() function that is called from both constructors to reduce code duplication or wait for c++11 so that we can use constructor delegation?

Maybe look through the commits and let me know what works for you and what doesn't - maybe you want this being approached completely differently? :)

Thanks and no rush!

Matti.

mattigruener commented on src/GafferBindings/CompoundNumericPlugBinding.cpp:
```diff
@@ -127,6 +182,15 @@ void bind()
 {
 	typedef typename T::ValueType V;
 
+	enum_<IECore::GeometricData::Interpretation>("Interpretation")
```
Yes, that's in there because had my own enum for a bit and then switched to the IECore one. Search/Replace "fixed" this one as well, but you're right - I need to get rid of it.

mattigruener commented:
Thanks for your review! I will try to address your requested changes as soon as possible :)

mattigruener commented:
> Can you elaborate a bit on what you're doing here?

Yes, so imagine a node that is supposed to let users create primitive variables in the UI and then provides a plug for users to specify the values of those primVars. Internally, that node sets up an `oslObject`, `oslOutObject` and then for each primVar creates an Osl shader that loads `ObjectProcessing/OutPoint`, for example. Determining which of the `Out*` shaders to load is easy when triggered by a user specifying through a menu that they want to create a point.

Now, on that node I have added a `PlugAdder` that creates that primVar once someone drags a plug on it. But now I don't have any way to differentiate between points, normals and vectors because all I'm seeing is that a user dragged a `V3fPlug` onto the `PlugAdder`. Does that make sense? That's why I need that interpretation to tell me which of the `Out*` shaders to load.

johnhaddon commented:
> Now, on that node I have added a PlugAdder that creates that primVar once someone drags a plug on it. But now I don't have any way to differentiate between points, normals and vectors because all I'm seeing is that a user dragged a V3fPlug onto the PlugAdder. Does that make sense? That's why I need that interpretation to tell me which of the Out* shaders to load.

Ah yes, makes perfect sense.

Overnight I think I've come round to this being the best approach, so you can ignore a lot of the waffle in my initial review comments. I've had another couple of thoughts though :

- I think `CompoundDataPlug::createPlugFromData()` needs updating to set the interpretation correctly.
- For the hashing, I was thinking that perhaps we should default the interpretation to None in the plug constructor, and then implement `CompoundNumericPlug::hash()` to only add it to the hash if it's non-None? That way, all the existing nodes which really don't care can remain unchanged and keep the same behaviour, and anything that does use the interpretation will have the hashing taken care of for them?
- We should probably update the plug addition menus in the OSLCode UI to add the option of point and normal plugs.
- We'll probably need to update the CompoundDataPlug menu in a similar way.

Cheers..
John

mattigruener commented on src/GafferBindings/CompoundNumericPlugBinding.cpp:
```diff
@@ -62,6 +64,50 @@ std::string compoundNumericPlugRepr( const T *plug )
+		if( const V3fPlug *v3fPlug = static_cast<const V3fPlug *>( plug ))
+		{
+			std::string interpretation;
+			switch( v3fPlug->getInterpretation() )
```
So, I've tried doing this by calling `IECorePython::repr` like so:

`
	std::string extraArgs = IECorePython::repr( interpretation );
`

I get a linker error suggesting that we don't have a repr() for `GeometricData::Interpretation` and indeed we couldn't find one. Am I doing it wrong or do we just need to define that `repr()`?

mattigruener commented on src/GafferOSL/OSLShader.cpp:
```diff
@@ -455,7 +455,27 @@ Plug *loadCompoundNumericParameter( const OSLQuery::Parameter *parameter, const
 	}
 
-	typename PlugType::Ptr plug = new PlugType( name, parent->direction(), defaultValue, minValue, maxValue, Plug::Default | Plug::Dynamic );
+	typename PlugType::Ptr plug;
```
I guess this is what you would want to move over to your `OpenImageIOAlgo.h`?

https://github.com/GafferHQ/gaffer/blob/master/src/GafferOSL/ShadingEngine.cpp#L98

mattigruener commented:
Hey John,

I just pushed code that should address most of your suggestions/feedback. I still have to do a couple of small things, I think, but I wanted to get this to you as soon as possible because I am sure that you will have a couple of things that you want changed. I hope I didn't misunderstand anything - especially the serialisation part is a little confusing... 

- Not sure if you want the menus to look like this?
- I didn't know how you want to differentiate between the color and non-color plugs in the `CompoundDataPlug` so I created `geometricCompoundNumericValuePlug`

I probably should have split that one mega-commit into multiple smaller ones. Let me know if you want me to do that before you look at it.

Thanks for you help with this :)

Matti.

johnhaddon commented on src/GafferBindings/CompoundNumericPlugBinding.cpp:
```diff
@@ -55,58 +55,59 @@ using namespace Gaffer;
-  public:
-
-	static std::string repr( const Gaffer::ValuePlug *plug, unsigned flagsMask, const std::string &extraArguments, const Serialisation *serialisation )
+	if( const V3fPlug *v3fPlug = IECore::runTimeCast<const V3fPlug>( plug ) ) 
```
This cast still isn't right, because it's preventing us from dealing with interpretations for V2fPlug etc. In fact we shouldn't need this cast at all - we know this function should only be instantiated for types T where T is a CompoundNumericPlug of some sort, so we should just be able to call `plug->interpretation()` and get on with it.

johnhaddon commented on src/GafferBindings/CompoundNumericPlugBinding.cpp:
```diff
@@ -55,58 +55,59 @@ using namespace Gaffer;
 	{
-		std::string result = ValuePlugSerialiser::repr( plug, flagsMask, extraArguments, serialisation );
-		if( const V3fPlug *v3fPlug = static_cast<const V3fPlug *>( plug ))
+		// TODO: we currently can't do this, but that would remove some of the uglyness below
```
`IECorePython::repr()` is a misleading oddity that I don't really remember the reason for. It sounds like it should call into python and do `repr( o )`, but instead it's actually a bunch of overloads implemented in pure C++, that give you the equivalent for whatever types it's implemented for. I think either it was for performance or because at the time we didn't know how to actually do it properly, but either way I think we'd be better to remove it.

This is what I was thinking of doing, but hadn't explained sufficiently. It calls into Python to do the equivalent of `repr( plug.interpretation() ` :

```
boost::python::object interpretationAsPythonObject( plug->interpretation() );
boost::python::object interpretationRepr = i.attr( "__repr__" )();
std::string s = boost::python::extract<std::string>( r );
```

Unfortunately this reveals that `repr( interpretation )` isn't doing the right thing, and returns `IECore._IECore.Interpretation.Point` rather than `IECore.GeometricData.Interpretation.Point`. I think though that'd I'd rather still take this approach and replace the prefix with the right thing, so that we're robust to any addition enum values that might be added in cortex in the future.

We should also omit adding it to the extraArgs entirely when `interpretation == GeometricData::None` (which is the default anyway), so we get a more readable serialisation where possible.

johnhaddon commented on src/GafferBindings/CompoundNumericPlugBinding.cpp:
```diff
@@ -55,58 +55,59 @@ using namespace Gaffer;
 	virtual std::string constructor( const Gaffer::GraphComponent *graphComponent, const Serialisation &serialisation ) const
 	{
-		return repr( static_cast<const ValuePlug *>( graphComponent ), Plug::All & ~Plug::ReadOnly, "", &serialisation );
+		printf("constructor() called...\n");
```
Looks like this is left over from debugging?

johnhaddon commented on src/GafferBindings/CompoundNumericPlugBinding.cpp:
```diff
@@ -55,58 +55,59 @@ using namespace Gaffer;
 	{
-		return repr( static_cast<const ValuePlug *>( graphComponent ), Plug::All & ~Plug::ReadOnly, "", &serialisation );
+		printf("constructor() called...\n");
+		return maskedCompoundNumericPlugRepr( static_cast<const ValuePlug *>( graphComponent ), Plug::All & ~Plug::ReadOnly );
```
This cast isn't right - `maskedCompoundNumericPlugRepr()` demands a CompoundNumericPlug of some type (once my suggested changes are made), so we should pass `static_cast<const T *>( graphComponent )`. The cast is safe because serialisers are only called with the types for which they are registered, and we register `CompoundNumericPlugSerialiser<T>` for type `T`.

johnhaddon commented on src/Gaffer/CompoundDataPlug.cpp:
```diff
@@ -452,6 +452,26 @@ ValuePlugPtr CompoundDataPlug::compoundNumericValuePlug( const std::string &name
 }
 
 template<typename T>
+ValuePlugPtr CompoundDataPlug::geometricCompoundNumericValuePlug( const std::string &name, Plug::Direction direction, unsigned flags, const T *value )
```
Looks good. Let's have a test case though.

mattigruener commented:
Alright, here is the new version.

- dropped the changes in the menu of UserPlugs. I agree with everything you said above.
- I believe I have added all the tests that you wanted?

Have a great weekend :)

mattigruener commented:
I don't think I'm causing the errors here, am I?

```
======================================================================

ERROR: testLoadNetworkFromVersion0_23 (GafferOSLTest.OSLShaderTest.OSLShaderTest)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/build/GafferHQ/gaffer/install/gaffer-0.31.0.0-linux/python/GafferOSLTest/OSLShaderTest.py", line 508, in testLoadNetworkFromVersion0_23

    s.load()

RuntimeError: Exception : Line 114 of /home/travis/build/GafferHQ/gaffer/install/gaffer-0.31.0.0-linux/python/GafferOSLTest/scripts/networkVersion-0.23.2.1.gfr : RuntimeError: Exception : Cannot set value for plug "p.x" except during computation.

======================================================================

ERROR: testSerialiation (GafferOSLTest.OSLShaderTest.OSLShaderTest)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/build/GafferHQ/gaffer/install/gaffer-0.31.0.0-linux/python/GafferOSLTest/OSLShaderTest.py", line 146, in testSerialiation

    script2.execute( script.serialise() )

RuntimeError: Exception : Line 23 : RuntimeError: Exception : Cannot set value for plug "c.r" except during computation.
```

johnhaddon commented on python/GafferOSLTest/shaders/version1.osl:
```diff
@@ -83,6 +83,7 @@ shader version1
 	float typeChanged2 = 2,
 	color typeChanged3 = 3,
 	string typeChanged4 = "four",
+  vector typeChanged5 = vector(0, 0, 0),
```
Indentation.

johnhaddon commented on python/GafferOSLTest/shaders/version2.osl:
```diff
@@ -74,6 +74,7 @@ shader version2
 	color typeChanged2 = 2,
 	float typeChanged3 = 3,
 	int typeChanged4 = 4,
+  normal typeChanged5 = normal(0, 0, 0),
```
Indentation.

johnhaddon commented on python/GafferUI/CompoundDataPlugValueWidget.py:
```diff
@@ -118,7 +118,10 @@ def __addMenuDefinition( self ) :
 		result.append( "/Add/V3i", { "command" : IECore.curry( Gaffer.WeakMethod( self.__addItem ), "", IECore.V3iData( IECore.V3i( 0 ) ) ) } )
 		result.append( "/Add/V2f", { "command" : IECore.curry( Gaffer.WeakMethod( self.__addItem ), "", IECore.V2fData( IECore.V2f( 0 ) ) ) } )
-		result.append( "/Add/V3f", { "command" : IECore.curry( Gaffer.WeakMethod( self.__addItem ), "", IECore.V3fData( IECore.V3f( 0 ) ) ) } )
+		result.append( "/Add/V3f ( Vector )", { "command" : IECore.curry( Gaffer.WeakMethod( self.__addItem ), "", IECore.V3fData( IECore.V3f( 0 ), IECore.GeometricData.Interpretation.Vector ) ) } )
```
Should we be exposing the point/vector/normal options for V2i/V2f/V3i as well? What do you think to having "V3f" be a submenu, with "Point", "Vector" and "Normal" child items and doing that for the others too?

mattigruener replied:
That sounds good, - I'll do that :) Seems that I keep thinking of this as only useful for the 3d variant...

johnhaddon commented on src/GafferBindings/CompoundNumericPlugBinding.cpp:
```diff
@@ -54,6 +56,23 @@ namespace
+	std::string extraArgs = "";
+
+	const IECore::GeometricData::Interpretation interpretation = plug->interpretation();
+	if( interpretation != IECore::GeometricData::None && interpretation != IECore::GeometricData::Color )
```
Why wouldn't we preserve `GeometricData::Color`? It'd be an odd choice for someone to make perhaps, but I don't see why we'd want the serialisation to silently throw it away.

mattigruener replied:
True.

My thinking was that we only allow `Vector`, `Point` and `Normal` from the UI anyway and I remembered you saying that we want to keep the serialisation minimal - so I added `Color` in that conditional to remove it from the serialisation of the `Color3fPlug` if I remember correctly.

johnhaddon replied:
Ah, I see. But won't most (all?) Color3fPlugs have the default interpretation in practice anyway, so they'll be omitting it already?

mattigruener replied:
By "default" you mean `IECore::GeometricData::None`? The ones created on OSLShaders will have the "correct" interpretation set. I just tried creating a oslBuildColor node and when I query the interpretation I get `IECore.GeometricData.Interpretation.Color` while not getting that in the serialisation. I am happy to remove it from the conditional if you're ok with that showing up in the serialisation :)

mattigruener replied:
Dan suggested to set the interpretation to None for Color3fPlugs explicitly. So maybe detect Color here [0] and set it to None?

[0] https://github.com/mattigruener/gaffer/blob/26fe8bc7510c18e4e6f8f72658e575cdfc5de655/src/GafferOSL/OSLShader.cpp#L450

johnhaddon replied:
Yep, I meant `GeometricData::None`. I think my preference is to map OSL color parameters to `Color3fPlug( interpretation = None )`, so that we can remove the special case from the serialisation code and still get a concise serialisation. We've already encoded the "colorness" in the type of the plug and the type of the data coming from it, so I don't see the need to double up. I might even argue that we shouldn't have made `Color` an interpretation at all, since it's not really geometric...but that can wait for another Cortex version...

danieldresser replied:
I'm not necessarily in favour of setting it to None, but I would say that all our Color3f plugs should probably have the same interpretation.  So either we should set it to None here, or we should somehow set the default to be Color on Color3f plugs ( Matti pointed out that the latter might be difficult because Color3f is just a typedef ).

mattigruener replied:
> I might even argue that we shouldn't have made Color an interpretation at all, since it's not really geometric...but that can wait for another Cortex version...

Yeah, so potentially we would even remove it from here [0] and have it return None/Numeric to begin with (or rename that function to not include "geometric" in the name)?

[0] https://github.com/mattigruener/gaffer/blob/26fe8bc7510c18e4e6f8f72658e575cdfc5de655/src/GafferImage/OpenImageIOAlgo.cpp#L75

johnhaddon replied:
Looks like you got the same suggestion in before I finished my reply :). If you wanted, we could make the default be GeometricData::Color for Color3f plug by adding a type traits class of some sort. Or we could argue that Color3fPlug shouldn't have an interpretation anyway, and we were wrong to make all CompoundNumericPlugs have interpretations, and instead introduce a separate template class to use for V3fPlug etc. I'm more inclined to go with Matti's proposed solution though - this feature was always going to be a bit esoteric, so I'd prefer it to have as little impact on the API as possible.

mattigruener replied:
My concern about this approach is that we potentially add that `Color` case back in at some point (because something else might want to use OpenImageIOAlgo and needs it) and that then changes the behaviour of our plugs. That would be a little odd, wouldn't it? Feels like blocking it explicitly as in my first suggestion would be more future-proof even if a little more ugly, but I'm happy to do whatever you consider the better solution :)

johnhaddon replied:
Sorry, I wasn't specific enough in my answer. I was advocating for your first solution, the one that tweaks the OSLShader loading code. As you say, we shouldn't second guess what other clients of OpenImageIOAlgo might want, so should leave OpenImageIOAlgo as it is...

mattigruener replied:
Great, will do. Thanks :)

mattigruener commented:
I have pushed another update. I had to create the PlugAlgo actually because there wasn't any yet (unless I'm blind). Is this what you had in mind?

Let's see if the test goes through this time... Rebasing this string of fixes will be an absolute joy :P

mattigruener commented:
Ok, still an error - I thought that explicitly importing Gaffer in that __init__ (last commit) would fix this, but it didn't. What do I need to do to get this symbol?

johnhaddon commented on src/GafferOSL/OSLShader.cpp:
```diff
@@ -468,19 +469,8 @@ Plug *loadCompoundNumericParameter( const OSLQuery::Parameter *parameter, const
 	}
 
 	typename PlugType::Ptr plug = new PlugType( name, parent->direction(), defaultValue, minValue, maxValue, Plug::Default | Plug::Dynamic, interpretation );
+	PlugAlgo::replacePlug( parent, plug );
```
Let's do this in a separate commit, and do it consistently in all the `load*Parameter` functions, so we can actually remove the `transformConnectionOrValue()` function entirely.

johnhaddon commented on src/Gaffer/PlugAlgo.cpp:
```diff
@@ -0,0 +1,81 @@
+	connections.push_back( c );
+}
+
+void replacePlug( Gaffer::GraphComponent *parent, PlugPtr plug )
```
Ah! I wasn't sure what your "What do I need to do to get this symbol?" comment was about, but having tried to compile your branch I presume it was a linking error where `replacePlug()` wasn't found? You'll either need to declare this here explicitly as `void PlugAlgo::replacePlug( Gaffer::GraphComponent *parent, PlugPtr plug )`, or scope it inside `namespace PlugAlgo {}`. Otherwise the compiler decides to generate a `replacePlug()` symbol in the global namespace rather than line this up with previously declared function in `namespace PlugAlgo`. You can see examples of each approach in SceneAlgo.cpp and StringAlgo.cpp respectively.

mattigruener replied:
Ah thanks :) When I mentioned the symbol I was referring to the travis error :) Will jump on all of these once I get to work - thanks for the review!

mattigruener commented:
Hey John,

updated the PR. I should have caught most of these myself, sorry. Thanks for clarifying some of the conventions you use :)

It would be really neat if we could remove all of the repeating blocks for checking existing plugs and their type and defaultValue in code like this

```
StringVectorDataPlug *existingPlug = parent->getChild<StringVectorDataPlug>( name );
if(	existingPlug && *existingPlug->defaultValue() == *defaultValueData )
{
	return existingPlug;
}

StringVectorDataPlugPtr plug = new StringVectorDataPlug( name, parent->direction(), defaultValueData, Plug::Default | Plug::Dynamic );

PlugAlgo::replacePlug( parent, plug );

```

and have that handled by the more general case in `PlugAlgo::replacePlug()`. But I guess we can't really currently. We could wrap `replacePlug` into a function that does those checks, but I assume we'd lose the early return that prevents unnecessary plug construction...

Let me know if there's anything else you want me to do on this one :)

mattigruener commented:
I had to wrap all calls to `PlugAlgo::replacePlug()` into `if( existingPlug ){}` blocks to prevent invalid calls to `setFrom()`. That's because `PlugAlgo::replacePlug()` doesn't check if the two plugs have the same type. I hope the tests run through now :)

I don't find this particularly pretty and it adds to my urge of refactoring some of this. How do you feel about the current code? If you feel the same way, I'd love to hear how you'd restructure this to make it a little nicer :)

mattigruener commented on src/GafferOSL/ShadingEngine.cpp:
```diff
@@ -95,25 +95,6 @@ struct TypeDescFromType<Color3f>
 	}
 };
 
-GeometricData::Interpretation geometricInterpretationFromVecSemantics( TypeDesc::VECSEMANTICS semantics )
```
Yes, exactly. I think 3 of the different travis setups complained about this. The others were fine with it.

mattigruener commented:
I will just leave this as is for now and maybe revisit at a later date. Thanks for your help with this one, John! I will rebase and clean it up a little now.

johnhaddon commented:
Good stuff! Thanks Matti!
